### PR TITLE
Take right size from tensor in SGD && add tests

### PR DESCRIPTION
**Pull request by RodionovF**

### CODE REVIEW

File: tests/unit/test_optimizer.cpp [hunk 1]
```diff
@@ -41,14 +41,49 @@ TEST(TestOptimizer, SgdStepBias) {
...
+    const float learning_rate = 0.1;
+
+    dw::Parameters m_params;
+    bool trainable = false;
+
+    for (size_t i = 0; i < weights.size(); ++i) {
+        dw::Tensor weight(dw::Shape{1});
+        weight.data()[0] = weights[i];
+
+        m_params.push_back(dw::Parameter(std::move(weight), trainable));
```
Comment (TolyaTalamanov) [inline tests/unit/test_optimizer.cpp:56]: I believe `params` would be enough, what's the purpose to use `m` prefix ?

Comment (TolyaTalamanov) [inline tests/unit/test_optimizer.cpp:56]: `emplace_back`

File: tests/unit/test_optimizer.cpp [hunk 2]
```diff
@@ -41,14 +41,49 @@ TEST(TestOptimizer, SgdStepBias) {
...
+    sgd.step();
+
+    trainable = false;
+
+    for (size_t i = 0; i < weights.size(); ++i) {
+        float expected_weight = weights[i];
+        if (trainable) {
+            expected_weight -= learning_rate * grads[i];
+        }
+        EXPECT_FLOAT_EQ(expected_weight, m_params[i].data().data()[0]);
```
Comment (TolyaTalamanov) [inline tests/unit/test_optimizer.cpp:74]: We have a special function for that

File: tests/unit/test_optimizer.cpp [hunk 3]
```diff
@@ -65,7 +100,41 @@ TEST(TestOptimizer, SgdStepMatrix) {
     }
 
     auto sgd = dw::optimizer::SGD(m_params, learning_rate);
+    sgd.step();
+
+    for (size_t i = 0; i < weights.size(); ++i) {
+        for (size_t j = 0; j < weights[i].size(); ++j) {
+            float expected_weight = weights[i][j] - learning_rate * grads[i][j];
+            EXPECT_FLOAT_EQ(expected_weight, m_params[i].data().data()[j]);
```
Comment (TolyaTalamanov) [inline tests/unit/test_optimizer.cpp:108]: The same

File: tests/unit/test_optimizer.cpp [hunk 4]
```diff
@@ -65,7 +100,41 @@ TEST(TestOptimizer, SgdStepMatrix) {
...
+    };
+    const float learning_rate = 0.01;
+
+    dw::Parameters m_params;
 
+    for (size_t i = 0; i < weights.size(); ++i) {
+        dw::Tensor weight(dw::Shape{2, 2, 1, 3});
+        std::copy(weights[i].begin(), weights[i].end(), weight.data());
+
+        m_params.push_back(dw::Parameter(std::move(weight), true));
```
Comment (TolyaTalamanov) [inline tests/unit/test_optimizer.cpp:131]: `emplace_back`

**Review by TolyaTalamanov — COMMENTED** [review]
Please avoid using `EXPECT_FLOAT_EQ`, we have a special function for that which will be extended to compare floats via `threshold` soon

**Comment by TolyaTalamanov** [comment]
BTW where is the `reference` implementation for SGD ?

File: tests/unit/test_optimizer.cpp [hunk 5]
```diff
@@ -11,67 +13,154 @@ TEST(TestOptimizer, SgdStepBias) {
...
 
-    dw::Parameters m_params;
+    dw::Parameters params;
+    dw::Parameters reference_params;
 
     for (size_t i = 0; i < weights.size(); ++i) {
         dw::Tensor weight(dw::Shape{1});
-        weight.data()[0] = weights[i];
+        dw::Tensor reference_weight(dw::Shape{1});
+        std::copy(weights.begin() + i, weights.begin() + i + 1, weight.data());
```
Comment (TolyaTalamanov) [inline tests/unit/test_optimizer.cpp:22]: copy the single element ?

File: tests/unit/test_optimizer.cpp [hunk 6]
```diff
@@ -11,67 +13,154 @@ TEST(TestOptimizer, SgdStepBias) {
...
-
+    auto sgd = dw::optimizer::SGD(params, learning_rate);
     sgd.step();
 
+    dw::reference::SGDStep(reference_params, learning_rate);
+
     for (size_t i = 0; i < weights.size(); ++i) {
-        float expected_weight = weights[i] - learning_rate * grads[i];
-        EXPECT_FLOAT_EQ(expected_weight, m_params[i].data().data()[0]);
+        dw::testutils::AssertTensorEqual(reference_params[i].data() ,params[i].data());
```
Comment (TolyaTalamanov) [inline tests/unit/test_optimizer.cpp:40]: comma alignment is broken

**Comment by TolyaTalamanov** [comment]
Ok, let's redesign tests like this:
```cpp

struct SGDOptimizerTest: public ::testing::Test {

    void init(const std::vector<dw::Shape>& shapes) {
        for (auto&& sh : shapes) {
            params.emplace_back(dw::Tensor(sh));
            expected.emplace_back(dw::Tensor(sh));
        
            dw::initiliazer::uniform(param.back().data());
            dw::initiliazer::uniform(param.back().grad());
            param.data().copyTo(expected.back().data());
            param.grad().copyTo(expected.back().grad());
        }
    }

    dw::Parameters params;
    dw::Parameters expected;
};

TEST_F(SGDOptimzerTest, TestName) {
    init({dw::Shape{16, 32}, dw::Shape{10}, dw::Shape{2, 3, 4}, dw::Shape{10, 15, 16, 4});
    float lr = 0.01;

    // Depworks
    dw::optimizer::SGD sgd(params, lr);
    sgd.step();
    
    // Reference
    dw::reference::SGDStep(expected, lr);

    for (int i = 0; i < params.size(); ++i) {
        dw::AsertTensorEqual(expected[i].data(), params[i].data());
    }
}
```

I believe that's would be enough.

**Comment by RodionovF** [comment]
> Ok, let's redesign tests like this:
> 
> ```c++
> struct SGDOptimizerTest: public ::testing::Test {
> 
>     void init(const std::vector<dw::Shape>& shapes) {
>         for (auto&& sh : shapes) {
>             params.emplace_back(dw::Tensor(sh));
>             expected.emplace_back(dw::Tensor(sh));
>         
>             dw::initiliazer::uniform(param.back().data());
>             dw::initiliazer::uniform(param.back().grad());
>             param.data().copyTo(expected.back().data());
>             param.grad().copyTo(expected.back().grad());
>         }
>     }
> 
>     dw::Parameters params;
>     dw::Parameters expected;
> };
> 
> TEST_F(SGDOptimzerTest, TestName) {
>     init({dw::Shape{16, 32}, dw::Shape{10}, dw::Shape{2, 3, 4}, dw::Shape{10, 15, 16, 4});
>     float lr = 0.01;
> 
>     // Depworks
>     dw::optimizer::SGD sgd(params, lr);
>     sgd.step();
>     
>     // Reference
>     dw::reference::SGDStep(expected, lr);
> 
>     for (int i = 0; i < params.size(); ++i) {
>         dw::AsertTensorEqual(expected[i].data(), params[i].data());
>     }
> }
> ```
> 
> I believe that's would be enough.

next lvl tests :100:

**Review by l-bat — APPROVED** [review]

**Comment by TolyaTalamanov** [comment]
LGTM :+1: 

Thanks!

**Pull request merged**